chore: support modular core

vladitasev opened the pull request:
The `OpenUI5Support.ts` module must be able to work with any version of OpenUI5/SAPUI5. The latest OpenUI5 version `1.116` introduces some breaking changes (f.e. `getThemeRoot` no longer available via the central configuration class).

Additionally, the version must be resolved synchronously, therefore I reverted to the old version parsing (and removed the usage of the official version util)

WIP

Background: `OpenUI5Support.init()` is expected to be called by the framework as early as possible. Therefore, it must preload all OpenUI5 modules that might be used anywhere else in the code. 

I'm looking for feedback especially on: 
 - `isLoaded` (later renamed to `OpenUI5Detected`): is this the right check to tell if OpenUI5 is on the window or not
 - `isModularCore` (later renamed to `isAtLeastVersion116`): is there a better/simpler check to determine this than checking the version?

codeworrior commented:
I'm not aware of **_breaking_** changes in UI5 1.116. We just suggest to use a `_new_` API that has less dependencies than the old `sap/ui/core/Configuration`. This should reduce the code for a UI5 app which uses web-component wrapper controls only.

codeworrior commented on packages/base/src/features/OpenUI5Support.ts:
```diff
@@ -29,48 +29,73 @@ type OpenUI5CoreConfiguration = {
+
 	static isLoaded() {
-		return !!getCore();
+		return typeof window.sap?.ui?.getCore === "function";
```
@loginger does this still fit as a criterion for the new bootstrap?

loginger replied:
No it will not fit with the new bootstrap as we do not expose the global sap.ui.getCore() anymore. IMO we must do a probing require 'sap.ui.require("sap/ui/core/Core")' instead. If the module is not yet required also check the global. This should be ok for both worlds.

codeworrior commented:
I see, thanks for the clarification. Indeed, some private methods have been removed. We should have made the usage of those methods clear (and should have marked them as restricted).

Still, I would prefer if @loginger takes a look at this change as there are several places where I'm in doubt (e.g. usage of `sap.ui.getCore()` as indicator for the loading state, usage of `attachInit` to register an init listener).

@petermuessig recently mentioned that he would like to simplify the coupling of OpenUI5Support and OpenUI5/SAPUI5 APIs . This totally makes sense to me, but requires more time. The API surface that you currently have to dependent on is too big. We should come up with something more compact.

loginger commented on packages/base/src/features/OpenUI5Support.ts:
```diff
@@ -29,48 +29,73 @@ type OpenUI5CoreConfiguration = {
-
-				core.attachInit(() => {
+			window.sap.ui.require(["sap/ui/core/Core"], (Core: OpenUI5Core) => {
+				Core.attachInit(() => {
```
We introduced a new Core.ready() function returning a Promise. You should use the ready instead of attachInit here. This will then work for both bootstraps.

vladitasev replied:
I couldn't get `Core.ready` to work with the old version (only with the 1.116 version) so I branched the logic.

loginger replied:
What exactly is not working? You could also pass a function (Core.ready(callback)) wich should behave exactly like attachInit...

codeworrior replied:
As I understood it, Vladi had issues on the _version_ dimension: 1.115 does not support `Core.ready()` yet.
In other words, `isModularCore()` in the code above checks for version 1.116, not for "is CoreX used".

vladitasev replied:
Indeed, I meant that `Core.ready` can only be used with the newest version, therefore I must keep the old code for 1.115 and earlier.

vladitasev reviewed (commented):
We still have reliance on a private function (`_get`). 

```ts
static getLocaleDataObject() {
		if (!OpenUI5Support.isLoaded()) {
			return;
		}

		const LocaleData = window.sap.ui.require("sap/ui/core/LocaleData") as LocaleData;

		if (OpenUI5Support.isModularCore()) {
			const Localization = window.sap.ui.require("sap/base/i18n/Localization") as Localization;
			return LocaleData.getInstance(Localization.getLanguageTag())._get();
		}

		const Core = window.sap.ui.require("sap/ui/core/Core") as OpenUI5Core;
		const config = Core.getConfiguration();
		return LocaleData.getInstance(config.getLocale())._get();
	}
```

We need the whole CLDR object, because one of the benefits of the OpenUI5Support feature is that if OpenUI5 is present, its CLDRs are used (no need to fetch them from our build assets).

```ts
// reuse OpenUI5 CLDR if present
	const openUI5Support = getFeature<typeof OpenUI5Support>("OpenUI5Support");
	if (openUI5Support) {
		const cldrContent = openUI5Support.getLocaleDataObject();
		if (cldrContent) {
			// only if openui5 actually returned valid content
			setLocaleData(localeId, cldrContent);
			return;
		}
	}
```

Therefore, we need a way to get the whole object and set it.

Could we get an official API for this?

vladitasev commented:
Hello, 

Based on the above discussion, I've renamed 2 methods to make things clearer. Now it looks like this:

```ts
class OpenUI5Support {
	static isAtLeastVersion116() {
		const version = window.sap.ui!.version as string;
		const parts = version.split(".");
		return parts && parts[1] && parseInt(parts[1]) >= 116;
	}

	static OpenUI5Detected() {
		return typeof window.sap?.ui?.require === "function";
	}
}
```
